File: src/commonTest/kotlin/com/scoreo/application/GetHeadToHeadUseCaseEloTest.kt
package com.scoreo.application

import com.scoreo.infrastructure.InMemoryGameTypeRepository
import com.scoreo.infrastructure.InMemoryMatchRepository
import com.scoreo.infrastructure.InMemoryPlayerRepository
import com.scoreo.domain.model.GameType
import com.scoreo.domain.model.Match
import com.scoreo.domain.model.Player
import com.scoreo.domain.model.PlayerScore
import com.scoreo.domain.model.WinCondition
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class GetHeadToHeadUseCaseEloTest {

    private val gt1 = GameType("gt1", "HIGHEST", WinCondition.HIGHEST_SCORE)
    private val gt2 = GameType("gt2", "LOWEST", WinCondition.LOWEST_SCORE)

    @Test
    fun `two players one match basic elo`() {
        val useCase = useCase(
            players("p1" to "Alice", "p2" to "Bob"),
            gameTypes(gt1),
            matches(Match("m1", 1000L, "gt1", listOf(PlayerScore("p1", 10), PlayerScore("p2", 5)))),
        )
        val result = useCase()
        assertEquals(1216, result.find { it.playerId == "p1" }?.elo)
        assertEquals(1184, result.find { it.playerId == "p2" }?.elo)
    }

    @Test
    fun `two players two matches elo reverses`() {
        val useCase = useCase(
            players("p1" to "Alice", "p2" to "Bob"),
            gameTypes(gt1),
            matches(
                Match("m1", 1000L, "gt1", listOf(PlayerScore("p1", 10), PlayerScore("p2", 5))),
                Match("m2", 2000L, "gt1", listOf(PlayerScore("p1", 3), PlayerScore("p2", 8))),
            ),
        )
        val result = useCase()
        assertTrue(result.find { it.playerId == "p2" }!!.elo > result.find { it.playerId == "p1" }!!.elo)
        assertEquals(2, result.size)
    }

    @Test
    fun `three players one winner elo normalized`() {
        val useCase = useCase(
            players("p1" to "Alice", "p2" to "Bob", "p3" to "Charlie"),
            gameTypes(gt1),
            matches(Match("m1", 1000L, "gt1", listOf(
                PlayerScore("p1", 10), PlayerScore("p2", 5), PlayerScore("p3", 3)
            ))),
        )
        val result = useCase()
        val alice = result.find { it.playerId == "p1" }!!
        val bob = result.find { it.playerId == "p2" }!!
        val charlie = result.find { it.playerId == "p3" }!!

        assertTrue(alice.elo > 1200)
        assertTrue(bob.elo < 1200)
        assertTrue(charlie.elo < 1200)
        assertTrue(result.all { it.elo == 1200 || it.playerId == "p1" } || alice.elo > bob.elo)
    }

    @Test
    fun `three players two winners`() {
        val useCase = useCase(
            players("p1" to "Alice", "p2" to "Bob", "p3" to "Charlie"),
            gameTypes(gt1),
            matches(Match("m1", 1000L, "gt1", listOf(
                PlayerScore("p1", 10), PlayerScore("p2", 10), PlayerScore("p3", 5)
            ))),
        )
        val result = useCase()
        val alice = result.find { it.playerId == "p1" }!!
        val bob = result.find { it.playerId == "p2" }!!
        val charlie = result.find { it.playerId == "p3" }!!

        assertEquals(1208, alice.elo)
        assertEquals(1208, bob.elo)
        assertEquals(1184, charlie.elo)
    }

    @Test
    fun `four players one winner elo normalized`() {
        val useCase = useCase(
            players("p1" to "A", "p2" to "B", "p3" to "C", "p4" to "D"),
            gameTypes(gt1),
            matches(Match("m1", 1000L, "gt1", listOf(
                PlayerScore("p1", 10), PlayerScore("p2", 5), PlayerScore("p3", 4), PlayerScore("p4", 3)
            ))),
        )
        val result = useCase()

        val winner = result.find { it.playerId == "p1" }!!
        assertTrue(winner.elo > 1208)
        assertTrue(winner.elo < 1220)
        result.filter { it.playerId != "p1" }.forEach { assertTrue(it.elo < 1200) }
    }

    @Test
    fun `ex-aequo match elo unchanged`() {
        val useCase = useCase(
            players("p1" to "Alice", "p2" to "Bob"),
            gameTypes(gt1),
            matches(Match("m1", 1000L, "gt1", listOf(
                PlayerScore("p1", 10), PlayerScore("p2", 10)
            ))),
        )
        val result = useCase()
        assertEquals(1200, result.find { it.playerId == "p1" }?.elo)
        assertEquals(1200, result.find { it.playerId == "p2" }?.elo)
    }

    @Test
    fun `lowest score win condition`() {
        val useCase = useCase(
            players("p1" to "Alice", "p2" to "Bob"),
            gameTypes(gt2),
            matches(Match("m1", 1000L, "gt2", listOf(
                PlayerScore("p1", 3), PlayerScore("p2", 10)
            ))),
        )
        val result = useCase()
        assertEquals(1216, result.find { it.playerId == "p1" }?.elo)
        assertEquals(1184, result.find { it.playerId == "p2" }?.elo)
    }

    @Test
    fun `manual win condition`() {
        val gt = GameType("gt3", "MANUAL", WinCondition.MANUAL)
        val useCase = useCase(
            players("p1" to "Alice", "p2" to "Bob"),
            gameTypes(gt),
            matches(Match("m1", 1000L, "gt3", listOf(
                PlayerScore("p1", 10), PlayerScore("p2", 5)
            ), manualWinners = listOf("p1"))),
        )
        val result = useCase()
        assertEquals(1216, result.find { it.playerId == "p1" }?.elo)
        assertEquals(1184, result.find { it.playerId == "p2" }?.elo)
    }

    @Test
    fun `players with zero wins and losses excluded from leaderboard`() {
        val useCase = useCase(
            players("p1" to "Alice", "p2" to "Bob", "p3" to "Charlie"),
            gameTypes(gt1),
            matches(Match("m1", 1000L, "gt1", listOf(PlayerScore("p1", 10), PlayerScore("p2", 5)))),
        )
        val result = useCase()
        assertEquals(2, result.size)
        assertTrue(result.none { it.playerId == "p3" })
    }

    @Test
    fun `elo floor does not go negative`() {
        val useCase = useCase(
            players("p1" to "Loser", "p2" to "Winner"),
            gameTypes(gt1),
            List(100) { i ->
                Match("m$i", i.toLong(), "gt1", listOf(
                    PlayerScore("p2", 10), PlayerScore("p1", 0)
                ))
            },
        )
        val result = useCase()
        assertTrue(result.find { it.playerId == "p1" }!!.elo >= 0)
    }

    private fun useCase(
        players: List<Player>,
        gameTypes: List<GameType>,
        matches: List<Match>,
    ): GetHeadToHeadUseCase {
        val pr = InMemoryPlayerRepository()
        players.forEach { pr.save(it) }
        val gtr = InMemoryGameTypeRepository()
        gameTypes.forEach { gtr.save(it) }
        val mr = InMemoryMatchRepository()
        matches.forEach { mr.save(it) }
        return GetHeadToHeadUseCase(mr, gtr, pr)
    }

    private fun players(vararg p: Pair<String, String>): List<Player> =
        p.map { (id, name) -> Player(id, name) }

    private fun gameTypes(vararg gt: GameType): List<GameType> = gt.toList()

    private fun matches(vararg m: Match): List<Match> = m.toList()
}

File: apps/scoreo/src/application/getHeadToHeadUseCaseElo.test.ts
import { describe, expect, it } from 'vitest'
import type { GameType } from '../domain/model/gameType'
import type { Match } from '../domain/model/match'
import type { Player } from '../domain/model/player'
import { InMemoryGameTypeRepository } from '../infrastructure/testing/inMemoryGameTypeRepository'
import { InMemoryMatchRepository } from '../infrastructure/testing/inMemoryMatchRepository'
import { InMemoryPlayerRepository } from '../infrastructure/testing/inMemoryPlayerRepository'
import { GetHeadToHeadUseCase } from './getHeadToHeadUseCase'

const gt1: GameType = {
  id: 'gt1',
  name: 'HIGHEST',
  winCondition: 'HIGHEST_SCORE',
  tieBreakRule: 'NONE',
  tieBreakCondition: 'HIGHEST_SCORE',
  tieBreakLabel: null,
  moduleId: null,
  active: true,
}
const gt2: GameType = {
  id: 'gt2',
  name: 'LOWEST',
  winCondition: 'LOWEST_SCORE',
  tieBreakRule: 'NONE',
  tieBreakCondition: 'HIGHEST_SCORE',
  tieBreakLabel: null,
  moduleId: null,
  active: true,
}

function players(...pairs: [string, string][]): Player[] {
  return pairs.map(([id, name]) => ({ id, name, active: true }))
}

function match(id: string, date: number, gameTypeId: string, playerScores: Match['playerScores'], overrides: Partial<Match> = {}): Match {
  return { id, date, gameTypeId, playerScores, manualWinners: [], secondaryPlayerScores: [], rounds: [], ...overrides }
}

function buildUseCase(playerList: Player[], gameTypes: GameType[], matches: Match[]): GetHeadToHeadUseCase {
  const pr = new InMemoryPlayerRepository()
  playerList.forEach((p) => pr.save(p))
  const gtr = new InMemoryGameTypeRepository()
  gameTypes.forEach((gt) => gtr.save(gt))
  const mr = new InMemoryMatchRepository()
  matches.forEach((m) => mr.save(m))
  return new GetHeadToHeadUseCase(mr, gtr, pr)
}

describe('GetHeadToHeadUseCase ELO', () => {
  it('two players one match basic elo', () => {
    const useCase = buildUseCase(
      players(['p1', 'Alice'], ['p2', 'Bob']),
      [gt1],
      [match('m1', 1000, 'gt1', [{ playerId: 'p1', score: 10 }, { playerId: 'p2', score: 5 }])],
    )
    const result = useCase.invoke()
    expect(result.find((r) => r.playerId === 'p1')?.elo).toBe(1216)
    expect(result.find((r) => r.playerId === 'p2')?.elo).toBe(1184)
  })

  it('two players two matches elo reverses', () => {
    const useCase = buildUseCase(
      players(['p1', 'Alice'], ['p2', 'Bob']),
      [gt1],
      [
        match('m1', 1000, 'gt1', [{ playerId: 'p1', score: 10 }, { playerId: 'p2', score: 5 }]),
        match('m2', 2000, 'gt1', [{ playerId: 'p1', score: 3 }, { playerId: 'p2', score: 8 }]),
      ],
    )
    const result = useCase.invoke()
    expect(result.find((r) => r.playerId === 'p2')!.elo).toBeGreaterThan(result.find((r) => r.playerId === 'p1')!.elo)
    expect(result).toHaveLength(2)
  })

  it('three players one winner elo normalized', () => {
    const useCase = buildUseCase(
      players(['p1', 'Alice'], ['p2', 'Bob'], ['p3', 'Charlie']),
      [gt1],
      [
        match('m1', 1000, 'gt1', [
          { playerId: 'p1', score: 10 },
          { playerId: 'p2', score: 5 },
          { playerId: 'p3', score: 3 },
        ]),
      ],
    )
    const result = useCase.invoke()
    const alice = result.find((r) => r.playerId === 'p1')!
    const bob = result.find((r) => r.playerId === 'p2')!
    const charlie = result.find((r) => r.playerId === 'p3')!

    expect(alice.elo).toBeGreaterThan(1200)
    expect(bob.elo).toBeLessThan(1200)
    expect(charlie.elo).toBeLessThan(1200)
    expect(alice.elo).toBeGreaterThan(bob.elo)
  })

  it('three players two winners', () => {
    const useCase = buildUseCase(
      players(['p1', 'Alice'], ['p2', 'Bob'], ['p3', 'Charlie']),
      [gt1],
      [
        match('m1', 1000, 'gt1', [
          { playerId: 'p1', score: 10 },
          { playerId: 'p2', score: 10 },
          { playerId: 'p3', score: 5 },
        ]),
      ],
    )
    const result = useCase.invoke()
    expect(result.find((r) => r.playerId === 'p1')?.elo).toBe(1208)
    expect(result.find((r) => r.playerId === 'p2')?.elo).toBe(1208)
    expect(result.find((r) => r.playerId === 'p3')?.elo).toBe(1184)
  })

  it('four players one winner elo normalized', () => {
    const useCase = buildUseCase(
      players(['p1', 'A'], ['p2', 'B'], ['p3', 'C'], ['p4', 'D']),
      [gt1],
      [
        match('m1', 1000, 'gt1', [
          { playerId: 'p1', score: 10 },
          { playerId: 'p2', score: 5 },
          { playerId: 'p3', score: 4 },
          { playerId: 'p4', score: 3 },
        ]),
      ],
    )
    const result = useCase.invoke()
    const winner = result.find((r) => r.playerId === 'p1')!
    expect(winner.elo).toBeGreaterThan(1208)
    expect(winner.elo).toBeLessThan(1220)
    result.filter((r) => r.playerId !== 'p1').forEach((r) => expect(r.elo).toBeLessThan(1200))
  })

  it('ex-aequo match elo unchanged', () => {
    const useCase = buildUseCase(
      players(['p1', 'Alice'], ['p2', 'Bob']),
      [gt1],
      [match('m1', 1000, 'gt1', [{ playerId: 'p1', score: 10 }, { playerId: 'p2', score: 10 }])],
    )
    const result = useCase.invoke()
    expect(result.find((r) => r.playerId === 'p1')?.elo).toBe(1200)
    expect(result.find((r) => r.playerId === 'p2')?.elo).toBe(1200)
  })

  it('lowest score win condition', () => {
    const useCase = buildUseCase(
      players(['p1', 'Alice'], ['p2', 'Bob']),
      [gt2],
      [match('m1', 1000, 'gt2', [{ playerId: 'p1', score: 3 }, { playerId: 'p2', score: 10 }])],
    )
    const result = useCase.invoke()
    expect(result.find((r) => r.playerId === 'p1')?.elo).toBe(1216)
    expect(result.find((r) => r.playerId === 'p2')?.elo).toBe(1184)
  })

  it('manual win condition', () => {
    const gt3: GameType = { ...gt1, id: 'gt3', name: 'MANUAL', winCondition: 'MANUAL' }
    const useCase = buildUseCase(
      players(['p1', 'Alice'], ['p2', 'Bob']),
      [gt3],
      [match('m1', 1000, 'gt3', [{ playerId: 'p1', score: 10 }, { playerId: 'p2', score: 5 }], { manualWinners: ['p1'] })],
    )
    const result = useCase.invoke()
    expect(result.find((r) => r.playerId === 'p1')?.elo).toBe(1216)
    expect(result.find((r) => r.playerId === 'p2')?.elo).toBe(1184)
  })

  it('players with zero wins and losses excluded from leaderboard', () => {
    const useCase = buildUseCase(
      players(['p1', 'Alice'], ['p2', 'Bob'], ['p3', 'Charlie']),
      [gt1],
      [match('m1', 1000, 'gt1', [{ playerId: 'p1', score: 10 }, { playerId: 'p2', score: 5 }])],
    )
    const result = useCase.invoke()
    expect(result).toHaveLength(2)
    expect(result.some((r) => r.playerId === 'p3')).toBe(false)
  })

  it('elo floor does not go negative', () => {
    const matches = Array.from({ length: 100 }, (_, i) =>
      match(`m${i}`, i, 'gt1', [
        { playerId: 'p2', score: 10 },
        { playerId: 'p1', score: 0 },
      ]),
    )
    const useCase = buildUseCase(players(['p1', 'Loser'], ['p2', 'Winner']), [gt1], matches)
    const result = useCase.invoke()
    expect(result.find((r) => r.playerId === 'p1')!.elo).toBeGreaterThanOrEqual(0)
  })
})
